File: step_30_eval/harness/tools.py
"""Step 29 - the job tools join the registry: bash_background, job_status,
job_wait and job_kill, in both TOOLS and TOOL_SCHEMAS. The task tool now
accepts a list of descriptions; its schema changed, its entry here did not.
The hook points from step 27 are unchanged.

TOOLS is what can run; TOOL_SCHEMAS is what the main agent is offered in act
mode. The browser tools are in the first and not the second: only the browse
subagent is offered them. The computer tools and the memory tools are in
both, and the MCP tools join both when their servers start.
"""

import json
import subprocess
from concurrent.futures import ThreadPoolExecutor

from . import browser, computer, history, hooks, jobs, memory, plan, sandbox
from .browse import BROWSE_SCHEMA, browse
from .permissions import check
from .skills import read_skill
from .subagent import TASK_SCHEMA, task
from .todos import TODO_SCHEMA, write_todos


def bash(command: str) -> str:
    """Run a shell command and return its combined stdout and stderr."""
    try:
        result = sandbox.run(command)
    except subprocess.TimeoutExpired as expired:
        # A slow command is the model's problem to work around, not a reason
        # to take the session down. Hand the failure back as a result.
        return f"Timed out after {expired.timeout}s and was killed. Narrow it down."
    return history.cap((result.stdout + result.stderr) or "(no output)")


def read_file(path: str) -> str:
    """Read a file and return its contents."""
    with open(path) as f:
        return history.cap(f.read())


def write_file(path: str, content: str) -> str:
    """Create a file, or overwrite it if it already exists."""
    with open(path, "w") as f:
        f.write(content)
    return f"Wrote {path}"


def str_replace(path, old_str, new_str, allow_multi_edit=False):
    """Swap exact text in a file. old_str must match exactly once."""
    with open(path) as f:
        content = f.read()

    count = content.count(old_str)
    if count == 0:
        return f"Error: old_str was not found in {path}"
    if count > 1 and not allow_multi_edit:
        return (
            f"Error: old_str matches {count} times in {path}. "
            "Add surrounding lines to make it unique, "
            "or set allow_multi_edit to replace them all."
        )

    with open(path, "w") as f:
        f.write(content.replace(old_str, new_str))
    return f"Replaced {count} match(es) in {path}"


TOOL_SCHEMAS = [
    {
        "type": "function",
        "function": {
            "name": "bash",
            "description": "Run a shell command and return its combined stdout and stderr.",
            "parameters": {
                "type": "object",
                "properties": {"command": {"type": "string", "description": "The shell command to run"}},
                "required": ["command"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "read_file",
            "description": "Read a file and return its contents.",
            "parameters": {
                "type": "object",
                "properties": {"path": {"type": "string", "description": "Path to the file to read"}},
                "required": ["path"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "read_skill",
            "description": "Open a skill by name and return its full instructions.",
            "parameters": {
                "type": "object",
                "properties": {"name": {"type": "string", "description": "Name of the skill to open"}},
                "required": ["name"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "write_file",
            "description": "Create a file, or overwrite it if it already exists.",
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {"type": "string", "description": "File to write"},
                    "content": {"type": "string", "description": "The full contents"},
                },
                "required": ["path", "content"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "str_replace",
            "description": (
                "Replace exact text in a file. old_str must appear exactly once, "
                "so include surrounding lines if needed."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "path": {"type": "string", "description": "File to edit"},
                    "old_str": {"type": "string", "description": "Exact text to find"},
                    "new_str": {"type": "string", "description": "Text to put in its place"},
                    "allow_multi_edit": {"type": "boolean", "description": "Replace every match instead of failing"},
                },
                "required": ["path", "old_str", "new_str"],
            },
        },
    },
    TODO_SCHEMA,
    TASK_SCHEMA,
    BROWSE_SCHEMA,
    *computer.COMPUTER_SCHEMAS,
    *memory.MEMORY_SCHEMAS,
    *jobs.JOB_SCHEMAS,
]

TOOLS = {
    "bash": bash,
    "read_file": read_file,
    "write_file": write_file,
    "str_replace": str_replace,
    "read_skill": read_skill,
    "write_todos": write_todos,
    "task": task,
    "browse": browse,
    "submit_plan": plan.submit_plan,  # runnable, but offered only in plan mode
    **browser.TOOLS,  # runnable, but offered only to the browse subagent
    **computer.COMPUTER_TOOLS,
    **memory.MEMORY_TOOLS,
    **jobs.JOB_TOOLS,
}


MAX_WORKERS = 4  # tool calls of one reply that may run at the same time

DENIED = "The user denied this tool call."


def decide(tool_call):
    """Parse the arguments and rate the call. Returns (args, action, reason).

    Nothing runs here. This is the half of execute() that must stay on the
    main thread, because an `ask` verdict turns into a prompt. The verdict
    is allow, ask or deny from the rules, or `blocked` from a PreToolUse
    hook. A denied call is not offered to the hooks: the rules said no first.
    """
    args = json.loads(tool_call.function.arguments)
    action, reason = check(tool_call.function.name, args)
    if action == "deny":
        return args, action, reason
    outcome = hooks.run_hooks("PreToolUse", {"tool_name": tool_call.function.name, "tool_input": args})
    if outcome.blocked:
        return args, "blocked", outcome.reason
    return args, action, reason


def run(tool_call, args):
    """Run the tool with already-parsed arguments, then the PostToolUse hooks.

    No permission check here. A hook that answers with a result replaces
    what the tool returned; the model sees the hook's version.
    """
    result = TOOLS[tool_call.function.name](**args)
    outcome = hooks.run_hooks("PostToolUse", {"tool_name": tool_call.function.name, "tool_input": args, "tool_result": result})
    if outcome.result is not None:
        return outcome.result if isinstance(outcome.result, str) else json.dumps(outcome.result)
    return result


def settle(action, reason):
    """Turn a verdict into a result string, or None when the call may run.

    A `deny` never runs, and neither does a call a hook `blocked`. An `ask`
    prompts the user and runs only on yes.
    """
    from .ui import ui  # here, not at the top: ui imports todos, tools imports ui

    if action == "deny":
        return f"Blocked by policy: {reason}"
    if action == "blocked":
        return f"Blocked by hook: {reason}"
    if action == "ask" and not ui.approve(reason):
        return DENIED
    return None


def execute(tool_call):
    """Run one tool call through the permission layer. Returns (args, result).

    Shared by the main loop and by subagents, so a subagent is fenced in by
    exactly the same rules - it is not a way around them. This is decide,
    settle and run in one step, for callers that want the direct path.
    """
    args, action, reason = decide(tool_call)
    result = settle(action, reason)
    if result is not None:
        return args, result
    return args, run(tool_call, args)


def execute_all(tool_calls):
    """Run every tool call of one reply. Returns [(args, result)] in the same order.

    One call takes the direct path. Several calls are decided first, one at a
    time on this thread, so the prompts appear in order. Then the allowed
    ones run together in a thread pool. A denied or declined call gets its
    message as the result and never runs.
    """
    if len(tool_calls) == 1:
        return [execute(tool_calls[0])]

    outcomes = []  # (args, result) per call; result is None until it has run
    for tool_call in tool_calls:
        args, action, reason = decide(tool_call)
        outcomes.append((args, settle(action, reason)))

    pending = [i for i, (_, result) in enumerate(outcomes) if result is None]
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
        futures = {i: pool.submit(run, tool_calls[i], outcomes[i][0]) for i in pending}
        for i, future in futures.items():
            outcomes[i] = (outcomes[i][0], future.result())
    return outcomes

File: step_45_typescript_core/harness-ts/tools.ts
/** Step 45 - the tools, and execute(), the one permission-checked entry point
 * the main loop and the subagent both use.
 *
 * Every tool takes one object of named arguments, parsed from the model's
 * JSON. Python spreads that dict as keyword arguments; here the object is
 * passed whole and destructured in the signature.
 */

import { readFileSync, writeFileSync } from "node:fs";

import * as history from "./history.ts";
import * as sandbox from "./sandbox.ts";
import { check } from "./permissions.ts";
import { readSkill } from "./skills.ts";
import { TASK_SCHEMA, task } from "./subagent.ts";
import { TODO_SCHEMA, writeTodos } from "./todos.ts";
import type { Args, Tool, ToolCall, ToolSchema } from "./types.ts";
import { ui } from "./ui.ts";

/** Run a shell command and return its combined stdout and stderr. */
export async function bash({ command }: { command: string }): Promise<string> {
  let result: sandbox.Output;
  try {
    result = await sandbox.run(command);
  } catch (failure) {
    if (failure instanceof sandbox.TimedOut) {
      // A slow command is the model's problem to work around, not a reason
      // to take the session down. Hand the failure back as a result.
      return `Timed out after ${failure.timeout}s and was killed. Narrow it down.`;
    }
    throw failure;
  }
  return history.cap(result.stdout + result.stderr || "(no output)");
}

/** Read a file and return its contents. */
export function readFile({ path }: { path: string }): string {
  return history.cap(readFileSync(path, "utf-8"));
}

/** Create a file, or overwrite it if it already exists. */
export function writeFile({ path, content }: { path: string; content: string }): string {
  writeFileSync(path, content, "utf-8");
  return `Wrote ${path}`;
}

/** Swap exact text in a file. old_str must match exactly once. */
export function strReplace({ path, old_str, new_str, allow_multi_edit = false }: {
  path: string; old_str: string; new_str: string; allow_multi_edit?: boolean;
}): string {
  const content = readFileSync(path, "utf-8");

  const pieces = content.split(old_str);
  const count = pieces.length - 1;
  if (count === 0) {
    return `Error: old_str was not found in ${path}`;
  }
  if (count > 1 && !allow_multi_edit) {
    return (
      `Error: old_str matches ${count} times in ${path}. ` +
      "Add surrounding lines to make it unique, " +
      "or set allow_multi_edit to replace them all."
    );
  }

  writeFileSync(path, pieces.join(new_str), "utf-8");
  return `Replaced ${count} match(es) in ${path}`;
}

export const TOOL_SCHEMAS: ToolSchema[] = [
  {
    type: "function",
    function: {
      name: "bash",
      description: "Run a shell command and return its combined stdout and stderr.",
      parameters: {
        type: "object",
        properties: { command: { type: "string", description: "The shell command to run" } },
        required: ["command"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "read_file",
      description: "Read a file and return its contents.",
      parameters: {
        type: "object",
        properties: { path: { type: "string", description: "Path to the file to read" } },
        required: ["path"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "read_skill",
      description: "Open a skill by name and return its full instructions.",
      parameters: {
        type: "object",
        properties: { name: { type: "string", description: "Name of the skill to open" } },
        required: ["name"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "write_file",
      description: "Create a file, or overwrite it if it already exists.",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string", description: "File to write" },
          content: { type: "string", description: "The full contents" },
        },
        required: ["path", "content"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "str_replace",
      description:
        "Replace exact text in a file. old_str must appear exactly once, " +
        "so include surrounding lines if needed.",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string", description: "File to edit" },
          old_str: { type: "string", description: "Exact text to find" },
          new_str: { type: "string", description: "Text to put in its place" },
          allow_multi_edit: { type: "boolean", description: "Replace every match instead of failing" },
        },
        required: ["path", "old_str", "new_str"],
      },
    },
  },
  TODO_SCHEMA,
  TASK_SCHEMA,
];

export const TOOLS: Record<string, Tool> = {
  bash,
  read_file: readFile,
  write_file: writeFile,
  str_replace: strReplace,
  read_skill: readSkill,
  write_todos: writeTodos,
  task,
};

/** Run one tool call through the permission layer. Returns [args, result].
 *
 * Shared by the main loop and by subagents, so a subagent is fenced in by
 * exactly the same rules - it is not a way around them.
 */
export async function execute(toolCall: ToolCall): Promise<[Args, string]> {
  const args: Args = JSON.parse(toolCall.function.arguments);
  const [action, reason] = check(toolCall.function.name, args);
  if (action === "deny") {
    return [args, `Blocked by policy: ${reason}`];
  }
  if (action === "ask" && !(await ui.approve(reason))) {
    return [args, "The user denied this tool call."];
  }
  return [args, await TOOLS[toolCall.function.name](args)];
}
